packages/cli/lib/base-generator.js

```diff
@@ -28,11 +32,235 @@ module.exports = class BaseGenerator extends Generator {
    * Subclasses can extend _setupGenerator() to set up the generator
    */
   _setupGenerator() {
+    this.option('config', {
+      type: String,
+      alias: 'c',
+      description: 'JSON file name or value to configure options',
+    });
+
+    this.option('yes', {
+      type: Boolean,
+      alias: 'y',
+      description:
+        'Skip all confirmation prompts with default or provided value',
+    });
+
     this.artifactInfo = this.artifactInfo || {
       rootDir: 'src',
     };
   }
 
+  /**
+   * Read a json document from stdin
+   */
+  _readJSONFromStdin() {
+    const rl = readline.createInterface({
+      input: process.stdin,
+    });
+
+    const lines = [];
+    if (process.stdin.isTTY) {
+      this.log(
+        chalk.green(
+          'Please type in a json object line by line ' +
+            '(Press <ctrl>-D or type EOF to end):',
+        ),
+      );
+    }
+
+    let err;
+    return new Promise((resolve, reject) => {
+      rl.on('SIGINT', () => {
+        err = new Error('Canceled by user');
+        rl.close();
+        reject(err);
+      })
+        .on('line', line => {
+          if (line === 'EOF') {
+            rl.close();
+          } else {
+            lines.push(line);
+          }
+        })
+        .on('close', () => {
+          if (err) return;
+          const jsonStr = lines.join('\n');
+          try {
+            const json = JSON.parse(jsonStr);
```

Comment:
Please move handling of the input outside of the code dealing with `stdin`.

Ideally, I'd like to see something like the following in `_readJSONFromStdin`:

```ts
async _readJSONFromStdin() {
  const input = await readFromStdin();
  try {
    return JSON.parse(input);
  } catch (err) {
    if (!process.stdin.isTTY) {
      debug(err, jsonStr);
    }
    throw err;
  }
}
```

Reply:
Fixed. Refactor the function into utils.